### DISCUSSION TITLE
Showing errors on blur with Select?

### DISCUSSION THREAD

**Discussion by lewisl9029** [ariakit/ariakit#1566] [Questions] [answered]
I've been trying to implement the standard show-errors-on-blur validation pattern with a select built on top of ariakit's Select.

My initial naive attempt involved just passing an onBlur to the Select, which unfortunately gets called as soon as the select is opened, not when we move focus away from the Select: https://codesandbox.io/s/infallible-bush-i0ll6m

I then looked around a bit and found [an example with the FormField component](https://ariakit.org/examples/form-select) that works around this issue using an onClose prop (that doesn't seem to be exposed on the Select for some reason? and I haven't been able to figure out where that's coming from or how it's hooked up). But this implementation has a different issue in that if you just tab over the select without opening/closing the popover, then the error doesn't get triggered: https://codesandbox.io/s/nostalgic-mccarthy-fo5qwj?file=/src/form-select.tsx

After a bunch of fiddling, I eventually arrived at the following solution using custom onBlur handlers on both the Select and SelectPopover to work around both of the issues above: https://codesandbox.io/s/confident-hellman-m6xhid?file=/src/select.tsx

I'm not super confident in the robustness though, especially since I had to hack around some DOM synchronization issues with a setTimeout. Curious if you have any thoughts on better ways to accomplish this?

**Comment by diegohaz** [comment, marked as answer]
> I then looked around a bit and found [an example with the FormField component](https://ariakit.org/examples/form-select) that works around this issue using an onClose prop (that doesn't seem to be exposed on the Select for some reason? and I haven't been able to figure out where that's coming from or how it's hooked up).

The Form Select example needs to be improved. But you'll find the `onClose` implementation in the `select.tsx` file:

<img width="811" alt="Screenshot of the form select page showing the code editor with the select.tsx tab selected and the code snippet below" src="https://user-images.githubusercontent.com/3068563/176579209-711b1206-cd10-4c6c-91c5-87dc02d480cd.png">

> But this implementation has a different issue in that if you just tab over the select without opening/closing the popover, then the error doesn't get triggered: https://codesandbox.io/s/nostalgic-mccarthy-fo5qwj?file=/src/form-select.tsx

Yeah! We have to change that to use blur. There's also an annoying bug when re-opening the popover due to the use of `<SelectPopover modal>` in my backlog.

> After a bunch of fiddling, I eventually arrived at the following solution using custom onBlur handlers on both the Select and SelectPopover to work around both of the issues above: https://codesandbox.io/s/confident-hellman-m6xhid?file=/src/select.tsx
> 
> I'm not super confidence in the robustness though, especially since I had to hack around some DOM synchronization issues with a setTimeout. Curious if you have any thoughts on better ways to accomplish this?

If your `SelectPopover` component doesn't use `modal` or `portal` props, you can wrap both `Select` and `SelectPopover` in a div and check the [`relatedTarget`](https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/relatedTarget) property:

```jsx
<div
  onBlur={(event) => {
    if (event.currentTarget.contains(event.relatedTarget)) return;
    touch();
  }}
>
  <Select />
  <SelectPopover />
</div>
```

If the `SelectPopover` component is rendered in a portal, it's a bit more verbose, but the same technique can be used:

```jsx
const selectRef = useRef(null);
const popoverRef = useRef(null);

const onBlur = (event) => {
  if (selectRef.current?.contains(event.relatedTarget)) return;
  if (popoverRef.current?.contains(event.relatedTarget)) return;
  touch();
}

<Select ref={selectRef} onBlur={onBlur} />
<SelectPopover portal ref={popoverRef} onBlur={onBlur} />
```

**Reply by diegohaz** [reply]
Created #1567

**Reply by lewisl9029** [reply]
TIL about `event.relatedTarget`! Just gave this a try in a non-portal'd select, and it seems to be working perfectly! 
This is so much simpler than the mess I had before. Thank you so much! :)
